backend/inspirehep/pidstore/providers/cnum.py

```diff
@@ -53,12 +55,41 @@ def next(cls, data):
 
         adjusted_date = datetime.strptime(opening_date, "%Y-%m-%d").strftime("%y-%m-%d")
         cnum = f"C{adjusted_date}"
-        cnums_count = PersistentIdentifier.query\
-            .filter(PersistentIdentifier.pid_value.like(f"{cnum}%"))\
-            .filter(PersistentIdentifier.pid_type == "cnum")\
-            .count()
 
-        if cnums_count > 0:
-            cnum = f"{cnum}.{cnums_count}"
+        retry_count = 0
+        try:
+            # Try to get postfix number from latest pid
+            last_cnum = (
+                PersistentIdentifier.query.filter(
+                    PersistentIdentifier.pid_value.like(f"{cnum}%")
+                )
+                .filter(PersistentIdentifier.pid_type == "cnum")
+                .order_by(PersistentIdentifier.created.desc())
+                .first()
+            )
+            last_postfix = int(last_cnum.pid_value.split(".")[-1])
+            postfix = last_postfix + 1
+        except (ValueError, AttributeError):
+            # If last pid do not have a postfix number it will throw ValueError
+            # If no results found then throws AttributeError
+            postfix = 0
+        while retry_count < current_app.config.get(
+            "MAX_CNUM_GENERATION_RETRY_COUNT", 10
+        ):
+            if postfix > 0:
+                full_cnum = f"{cnum}.{postfix}"
+            else:
+                full_cnum = cnum
+            retry_count += 1
+            cnum_count = (
+                PersistentIdentifier.query.filter(
+                    PersistentIdentifier.pid_value.like(full_cnum)
+                )
+                .filter(PersistentIdentifier.pid_type == "cnum")
+                .count()
+            )
+            if cnum_count == 0:
+                break
+            postfix += 1
 
-        return cnum
+        return full_cnum
```

Comment:
IMO we should get the latest prefix and + 1 instead of this retry

Reply:
He added this retry because, apparently, we can have a cnum with prefix 2 created before a cnum with prefix 1 (which I don't understand why). But I also agree that this retry is ugly. I think we should query and find the biggest prefix and then do +1, the latest prefix won't work

Reply:
Latest prefix is not correct as many CNUMS were created in weird order. 
Check: 
SELECT * from pidstore_pid where pid_type='cnum' ORDER BY pid_value LIMIT 1000
"2020-02-21 12:07:05.623265"	"2020-02-21 12:07:05.62327"	22160560	"cnum"	"C00-01-04"
"2020-02-21 12:06:45.515772"	"2020-02-21 12:06:45.515778"	22157023	"cnum"	"C00-01-04.1"

So here if we would take latest cnum it would return "C00-01-04.1" as the next one.

Reply:
I think you need `DESC` also, but there's the issue that ` 10 < 2`, so it's not exactly lexicographic ordering :( if you want to do it in DB, you need to split on `.`, cast to `int` and order based on that. Or you can get everything in Python and compute the next one. In any case there might be (very improbable) collisions in case two concurrent transactions try to mint the same CNUM. The proper way to do this is probably to have another table that maps CNUM prefix to latest CNUM suffix and increment the counter when minting a new CNUM :cry:

Reply:
I added row locks for CNUM